test/test_util/test_data.py

```diff
@@ -0,0 +1,233 @@
+    if sample_size < 1:
+        assert size(X_sampled) == int(sample_size * size(X))
+    else:
+        assert size(X_sampled) == sample_size
+
+
+@parametrize('X', [np.asarray([[0, 0, 1]] * 10)])
+@parametrize('dtype', supported_precision_reductions + [np.dtype('float32'), np.dtype('float64')])
+@parametrize('x_type', [np.ndarray, csr_matrix])
+def test_reduce_precision_correctly_reduces_precision(X, dtype, x_type):
+    X = X.astype(dtype)
+    if x_type == csr_matrix:
+        X = x_type(X)
+
+    X_reduced, precision = reduce_precision(X)
+
+    # Check the reduced precision is correctly returned
+    assert X_reduced.dtype == precision
+
+    # Check that it was reduce to the correct precision
+    assert precision == reduction_mapping[dtype]
+
+    # Check that X's shape was not modified in any way
+    assert X.shape == X_reduced.shape
+
+    # Check that the return type is the one as we passed in
+    assert type(X) == type(X_reduced)
+
+
+@parametrize('X', [np.asarray([0, 0, 1]) * 10])
+@parametrize('dtype', [np.int32, np.int64, np.complex128])
+def test_reduce_precision_with_unsupported_dtypes(X, dtype):
+    X = X.astype(dtype)
+    with pytest.raises(ValueError):
+        reduce_precision(X)
+
+
+@parametrize("X", [
+    np.asarray([[1] * 10] * 100000, dtype=np.float64)  # Make it big for reductions to take place
+])
+@parametrize("x_type", [csr_matrix, np.ndarray])
+@parametrize("dtype", supported_precision_reductions)
+@parametrize('y, is_classification', [
+    (np.asarray([1] * 100000), True),
+    (np.asarray([1.0] * 100000), False),
+])
+@parametrize('multiplier', [1, 5.2, 10])
+@parametrize('operations', [['precision'], ['subsample'], ['precision', 'subsample']])
+def test_reduce_dataset_reduces_size_and_precision(
+    X, x_type, dtype, y, is_classification, multiplier, operations
+):
+    assert len(X) == len(y)
+    X = X.astype(dtype)
+    if x_type == csr_matrix:
+        X = x_type(X)
+
+    random_state = 0
+    memory_limit = 1  # Force reductions
+
+    X_out, y_out = reduce_dataset_size_if_too_large(
+        X=X,
+        y=y,
+        random_state=random_state,
+        memory_limit=memory_limit,
+        operations=operations,
+        multiplier=multiplier,
+        is_classification=is_classification,
+    )
+
+    def bytes(arr):
+        return arr.nbytes if isinstance(arr, np.ndarray) else arr.data.nbytes
+
+    # If we expect some precision reduction unless at float32 already
+    if 'precision' in operations and dtype != np.float32:
+        expected = reduction_mapping[X.dtype]
+        assert X_out.dtype == expected
+        assert bytes(X_out) < bytes(X)
+
+    # If we expect some subsampling
+    if 'subsample' in operations:
+        assert X_out.shape[0] < X.shape[0]
+        assert y_out.shape[0] < y.shape[0]
+        assert bytes(X_out) < bytes(X)
+
+
+def test_reduce_dataset_invalid_dtype_for_precision_reduction():
+    X = np.asarray([1, 2, 3], dtype=int)
+    with pytest.raises(ValueError):
```

Comment:
Could you please check for the error message here, too? Otherwise, it'll be a bit hard to review and later on know what to expect.

Reply:
Donee